### PR TITLE
ISPN-10275 add memcached TOUCH operation

### PR DESCRIPTION
**Pull request by brzler**
https://issues.jboss.org/browse/ISPN-10275

This PR will add support for the Memcached TOUCH operation.

### CODE REVIEW

**Comment by ryanemerson** [comment]
test this please

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java [hunk 1]
```diff
@@ -296,6 +301,26 @@ private void decodeValue(ChannelHandlerContext ctx, ByteBuf buffer, MemcachedDec
       writeResponse(ch, ret);
    }
 
+   private Object touch() {
+      final CacheEntry<byte[], byte[]> cacheEntry = cache.getCacheEntry(key);
+      if (cacheEntry == null) {
+         return createNotExistResponse();
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: It looks like this should return "NOT_FOUND\r\n" instead.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java [hunk 2]
```diff
@@ -296,6 +301,26 @@ private void decodeValue(ChannelHandlerContext ctx, ByteBuf buffer, MemcachedDec
...
+      }
+      final Metadata newMetadata = new MemcachedMetadata.Builder()
+              .merge(cacheEntry.getMetadata())
+              .lifespan(params.lifespan > 0 ? toMillis(params.lifespan) : -1)
+              .build();
+      cache.replace(key, cacheEntry.getValue(), newMetadata);
```
Comment (wburns) [inline server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: It is unclear from https://github.com/memcached/memcached/blob/master/doc/protocol.txt but we technically have 2 expiration types, lifespan and max idle.  It sounds like memcached uses the latter only. In this case we don't need to do a replace, but rather just the get alone would be sufficient to refresh the max idle time.

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java (hunk 1 above)
Comment (brzler) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: createNotExistResponse() actually returns a NOT_FOUND message ... I haven't named this already existing method :)

Comment (wburns) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: Great! Sorry, I didn't look more closely :)

File: server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java (hunk 2 above)
Comment (brzler) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: so infinispan is simply ignoring the exptime from the memcached protocol and also ignores the lifespan stored in the cache entry's metadata?
what else is the metadata->lifespan property used for and why is it stored with the key/value entry?

Comment (wburns) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: Sorry, I am rereading the protocol txt file you linked in the JIRA. It sounds like emulating the guarantees they state would be best using lifespan, and ignore max-idle as our put and your new code is doing.

Comment (brzler) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: so I did everything right by making a replace with new metadata (and lifespan)?

Comment (wburns) [reply server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java]: I am not sure what the semantics are around a concurrent touch with another write operation. But I would say we probably don't have to worry about it, as they are most likely undefined.

**Comment by wburns** [comment]
The only other thing I see looking at the document is whether we also need get and touch `gat` operation as well?

**Comment by brzler** [comment]
> The only other thing I see looking at the document is whether we also need get and touch `gat` operation as well?

```
java.lang.UnsupportedOperationException: Get and touch is not supported for ASCII protocol

	at net.spy.memcached.protocol.ascii.AsciiOperationFactory.getAndTouch(AsciiOperationFactory.java:86)
	at net.spy.memcached.MemcachedClient.asyncGetAndTouch(MemcachedClient.java:1510)
	at net.spy.memcached.MemcachedClient.getAndTouch(MemcachedClient.java:1167)
	...
```
Should I really implement features the used testing library isn't capable of?

**Comment by tristantarrant** [comment]
We could use xmemcached which does have that API.

**Comment by tristantarrant** [comment]
sorry, I realize now that we should really be implementing the binary protocol

**Comment by brzler** [comment]
so can this PR be merged or are there any other changes required?

**Comment by tristantarrant** [comment]
Merged, thanks !

**Pull request merged**